File: Medlem3060/AppEngine/medlem3060/medlem3060/clsInfotekst.py
# coding=utf-8 
from google.appengine.ext import db 
import re
import logging
from models import UserGroup, User, NrSerie, Kreditor, Kontingent, Tilpbs, Fak, Sftp, Infotekst, Sysinfo, Menu, MenuMenuLink, Medlemlog, Person

class clsInfotekstParam: 
  pass

class clsInfotekst(object):
  def __init__(self, p):
    if p.infotekst_id:
      self.infotekst_id =  p.infotekst_id
    if p.numofcol:
      self.numofcol = p.numofcol
    if p.navn_medlem:
      self.navn_medlem = p.navn_medlem
    if p.kaldenavn:
      self.kaldenavn = p.kaldenavn
    if p.fradato:
      self.fradato = p.fradato
    if p.tildato:
      self.tildato = p.tildato
    if p.betalingsdato:
      self.betalingsdato = p.betalingsdato
    if p.advisbelob:
      self.advisbelob = p.advisbelob
    if p.ocrstring:
      self.ocrstring = p.ocrstring
    if p.underskrift_navn:
      self.underskrift_navn = p.underskrift_navn
    if p.bankkonto:
      self.bankkonto = p.bankkonto
    if p.advistekst:
      self.advistekst = p.advistekst


  def delegate(self, match):
    v = '%s' % (match.group())
    vl = v.lower()
    if vl == "##navn_medlem##":
      if self.navn_medlem: 
        return self.navn_medlem
    elif vl ==  "##kaldenavn##":
      if self.kaldenavn:
        return self.kaldenavn
      if self.navn_medlem:
        return self.navn_medlem
    elif vl ==  "##fradato##":
      if self.fradato: 
         return self.fradato.strftime("%d. %B %Y")
    elif vl ==  "##tildato##":
      if self.tildato: 
        return self.tildato.strftime("%d. %B %Y")
    elif vl ==  "##betalingsdato##":
      if self.betalingsdato: 
        return self.betalingsdato.strftime("%A") + " den " + self.betalingsdato.strftime("%d. %B")
    elif vl ==  "##advisbelob##":
      if self.advisbelob:
        advisbelob_formated =  '%.2f' % (self.advisbelob)
        return advisbelob_formated.replace('.', ',')
    elif vl ==  "##ocrstring##":
      if self.ocrstring:
        return self.ocrstring
    elif vl ==  "##underskrift_navn##":
      if self.underskrift_navn:
        return self.underskrift_navn
    elif vl ==  "##bankkonto##":
      if self.bankkonto:
        return self.bankkonto
    elif vl ==  "##advistekst##":
      if self.advistekst: 
        return self.advistekst
    else:
      return v


  def getinfotekst(self):
    crlf = "\r\n"
    infotext = None
    try:
      keyInfotekst = db.Key.from_path('Persons','root','Infotekst','%s' % (self.infotekst_id))
      recInfotekst = Infotekst.get(keyInfotekst)
      infotext = recInfotekst.Msgtext
    except:
      infotext = None

    #RegexOptions options = ((RegexOptions.IgnorePatternWhitespace | RegexOptions.Multiline) | RegexOptions.IgnoreCase)
    #Regex regexParam = new Regex(@"(\#\#[^\#]+\#\#)", options)
    options = re.VERBOSE | re.MULTILINE | re.IGNORECASE
    p = re.compile(r'(\#\#[^\#]+\#\#)', options)

    MsgtextSub = ""
    if infotext:
      MsgtextSub = p.sub(self.delegate, infotext)
      if not self.numofcol:
        return MsgtextSub
      else:
        return self.splittextincolumns(MsgtextSub, self.numofcol)
    return ""

  
  def splittextincolumns(self, msg, numofcol):
    return self.formattext(msg, numofcol)


  def formattext(self, inputtext, maxlinewith):
    currentlinelength = 0
    outputtext = ""
    workline = ""
    linecount = 0
    wordcount = 0
    #crlf = "\r\n"
    crlf = "\n"
    bltp = ' '

    inputtextlines = inputtext.split(crlf)
    for currentline in inputtextlines:
      if len(currentline) <= maxlinewith:
        outputtext += currentline + "\r\n"
        linecount += 1
      else:
        currentlinelength = 0
        wordcount = 0
        currentlines = currentline.split(bltp)
        for currentword in currentlines:
          if len(currentword) > 0:
            if currentlinelength == 0:
              workline += currentword
              currentlinelength = len(currentword)
              wordcount += 1
            else:
              if (currentlinelength + 1 + len(currentword)) <= maxlinewith:
                workline += " " + currentword
                currentlinelength += 1 + len(currentword)
                wordcount += 1
              else:
                outputtext += self.expandline(workline, wordcount, maxlinewith) + "\r\n"
                linecount += 1
                workline = currentword
                currentlinelength = len(currentword)
                wordcount = 1

        if currentlinelength > 0:
          outputtext += workline + "\r\n"
          workline = ""
          linecount += 1

    return outputtext

       
  def expandline(self, inputline, wordsinline, outputlinewith):
    outputline = ""
    firstword = True
    splitchar = ' '

    blanksmissingcount = outputlinewith - len(inputline)
    blankscount = abs(blanksmissingcount / (wordsinline - 1))
    blankscountmodulo = blanksmissingcount % (wordsinline - 1)
    inputwords = inputline.split(splitchar)
    for inputword in inputwords:
      if firstword:
        outputline = inputword
      else:
        if blankscountmodulo > 0:
          outputline += inputword.rjust(len(inputword) + blankscount + 2, ' ')
          blankscountmodulo -= 1
        else:
          outputline += inputword.rjust(len(inputword) + blankscount + 1, ' ')
      firstword = False
 
    return outputline

File: _Puls3060/Puls3060Medlem3060/Pbs3060/clsInfotekst.cs

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
//using System.Drawing;
using System.Linq;
using System.Text;
//using System.Windows.Forms;
using Microsoft.Win32;
using System.IO;
using Microsoft.VisualBasic;
//using System.Transactions;
using System.Text.RegularExpressions;
using MimeKit;

namespace nsPbs3060
{
    public class clsInfotekst
    {
        public int? infotekst_id { get; set; }
        public int? numofcol { get; set; }
        public string navn_medlem { get; set; }
        public string kaldenavn { get; set; }
        public DateTime? fradato { get; set; }
        public DateTime? tildato { get; set; }
        public DateTime? betalingsdato { get; set; }
        public decimal? advisbelob { get; set; }
        public string ocrstring { get; set; }
        public string underskrift_navn { get; set; }
        public string bankkonto { get; set; }
        public string advistekst { get; set; }
        public string sendtsom { get; set; }
        public string kundenr { get; set; }

        public string getinfotekst(dbData3060DataContext dbData3060)
        {
            string[] crlf = { "\r\n" };
            string infotext = null;

            try { infotext = (from i in dbData3060.tblinfoteksts where i.id == infotekst_id select i).First().msgtext; }
            catch (System.InvalidOperationException) { }

            RegexOptions options = ((RegexOptions.IgnorePatternWhitespace | RegexOptions.Multiline) | RegexOptions.IgnoreCase);
            Regex regexParam = new Regex(@"(\#\#[^\#]+\#\#)", options);

            string MsgtextSub = "";
            if (infotext != null)
            {
                MsgtextSub = regexParam.Replace(infotext, delegate(Match match)
                {
                    string v = match.ToString();
                    switch (v.ToLower())
                    {
                        case "##navn_medlem##":
                            if (navn_medlem != null) return navn_medlem;
                            break;

                        case "##kaldenavn##":
                            if (kaldenavn != null) return kaldenavn;
                            if (navn_medlem != null) return navn_medlem;
                            break;

                        case "##fradato##":
                            if (fradato != null) return string.Format("{0:d. MMMM yyyy}", fradato);
                            break;

                        case "##tildato##":
                            if (tildato != null) return string.Format("{0:d. MMMM yyyy}", tildato);
                            break;

                        case "##betalingsdato##":
                            if (betalingsdato != null) return string.Format("{0:dddd}", betalingsdato) + " den " + string.Format("{0:d. MMMM}", betalingsdato);
                            break;

                        case "##advisbelob##":
                            if (advisbelob != null) return String.Format("{0:###0.00}", advisbelob).Replace('.', ',');
                            break;

                        case "##ocrstring##":
                            if (ocrstring != null) return ocrstring;
                            break;

                        case "##underskrift_navn##":
                            if (underskrift_navn != null) return underskrift_navn;
                            break;

                        case "##bankkonto##":
                            if (bankkonto != null) return bankkonto;
                            break;

                        case "##advistekst##":
                            if (advistekst != null) return advistekst;
                            break;

                        case "##sendtsom##":
                            if (sendtsom != null) return sendtsom;
                            break;

                        case "##kundenr##":
                            if (sendtsom != null) return kundenr;
                            break;
                    }
                    return v;
                });

                if (numofcol == null)
                {
                    return MsgtextSub;
                }
                else
                {
                    return splittextincolumns(MsgtextSub, (int)numofcol);
                }
            }
            return "";
        }

        public string splittextincolumns(string msg, int numofcol)
        {
            return formattext(msg, numofcol);
        }

        public string formattext(string inputtext, int maxlinewith)
        {

            int currentlinelength;
            string outputtext = "";
            string workline = "";
            int linecount = 0;
            int wordcount = 0;
            string[] crlf = { "\r\n", "\n" };
            char[] bltp = { '\t', ' ' };

            string[] inputtextlines = inputtext.Split(crlf, StringSplitOptions.None);
            foreach (string currentline in inputtextlines)
            {
                if (currentline.Length <= maxlinewith)
                {
                    outputtext += currentline + "\r\n";
                    linecount++;
                }
                else
                {
                    currentlinelength = 0;
                    wordcount = 0;
                    string[] currentlines = currentline.Split(bltp);
                    foreach (string currentword in currentlines)
                    {
                        if (currentword.Length > 0)
                        {
                            if (currentlinelength == 0)
                            {
                                workline += currentword;
                                currentlinelength = currentword.Length;
                                wordcount++;
                            }
                            else
                            {
                                if ((currentlinelength + 1 + currentword.Length) <= maxlinewith)
                                {
                                    workline += " " + currentword;
                                    currentlinelength += 1 + currentword.Length;
                                    wordcount++;
                                }
                                else
                                {
                                    outputtext += expandline(workline, wordcount, maxlinewith) + "\r\n";
                                    linecount++;
                                    workline = currentword;
                                    currentlinelength = currentword.Length;
                                    wordcount = 1;
                                }
                            }
                        }
                    }

                    if (currentlinelength > 0)
                    {
                        outputtext += workline + "\r\n";
                        workline = "";
                        linecount++;
                    }
                }
            }
            return outputtext;
        }

        public string expandline(string inputline, int wordsinline, int outputlinewith)
        {
            string outputline = "";
            bool firstword = true;
            char[] splitchar = { ' ' };

            int blanksmissingcount = outputlinewith - inputline.Length;
            int blankscount = Math.Abs(blanksmissingcount / (wordsinline - 1));
            int blankscountmodulo = blanksmissingcount % (wordsinline - 1);
            string[] inputwords = inputline.Split(splitchar);
            foreach (string inputword in inputwords)
            {
                if (firstword)
                {
                    outputline = inputword;
                }
                else
                {
                    if (blankscountmodulo > 0)
                    {
                        outputline += inputword.PadLeft(inputword.Length + blankscount + 2, ' ');
                        blankscountmodulo--;
                    }
                    else
                    {
                        outputline += inputword.PadLeft(inputword.Length + blankscount + 1, ' ');
                    }
                }
                firstword = false;
            }
            return outputline;
        }

        public string substitute_message(string infotext)
        {
            RegexOptions options = ((RegexOptions.IgnorePatternWhitespace | RegexOptions.Multiline) | RegexOptions.IgnoreCase);
            Regex regexParam = new Regex(@"(\#\#[^\#]+\#\#)", options);

            string MsgtextSub = "";
            if (infotext != null)
            {
                MsgtextSub = regexParam.Replace(infotext, delegate(Match match)
                {
                    string v = match.ToString();
                    switch (v.ToLower())
                    {
                        case "##navn_medlem##":
                            if (navn_medlem != null) return navn_medlem;
                            break;

                        case "##kaldenavn##":
                            if (kaldenavn != null) return kaldenavn;
                            if (navn_medlem != null) return navn_medlem;
                            break;

                        case "##fradato##":
                            if (fradato != null) return string.Format("{0:d. MMMM yyyy}", fradato);
                            break;

                        case "##tildato##":
                            if (tildato != null) return string.Format("{0:d. MMMM yyyy}", tildato);
                            break;

                        case "##betalingsdato##":
                            if (betalingsdato != null) return string.Format("{0:dddd}", betalingsdato) + " den " + string.Format("{0:d. MMMM}", betalingsdato);
                            break;

                        case "##advisbelob##":
                            if (advisbelob != null) return String.Format("{0:###0.00}", advisbelob).Replace('.', ',');
                            break;

                        case "##ocrstring##":
                            if (ocrstring != null) return ocrstring;
                            break;

                        case "##underskrift_navn##":
                            if (underskrift_navn != null) return underskrift_navn;
                            break;

                        case "##bankkonto##":
                            if (bankkonto != null) return bankkonto;
                            break;

                        case "##advistekst##":
                            if (advistekst != null) return advistekst;
                            break;

                        case "##sendtsom##":
                            if (sendtsom != null) return sendtsom;
                            break;

                        case "##kundenr##":
                            if (sendtsom != null) return kundenr;
                            break;
                    }
                    return v;
                });
            }
            return MsgtextSub;
        }
    }
}
